### ISSUE TITLE
Compile against recent google WebRTC snapshot for ICE support

### ISSUE THREAD

**Issue by fletcherdvalve** [ValveSoftware/GameNetworkingSockets#117]
We have ICE support working through google webrtc, but I can't remember when we grabbed the snapshot of their code.

See [README_P2P.md](https://github.com/ValveSoftware/GameNetworkingSockets/blob/master/README_P2P.md)

Tasks:
* Download a recent webrtc release, and get it building on at least one platform
* Setup cmake files (we use a combination of manually edited project files and scripts)
* Apply any code changes need to build steamwebrtc code (hopefully none!)
* Document how to build their stuff and what revision was downloaded
* Fix the other platforms people care about

**Comment by ZCube** [comment]
Maybe these scripts will help you.
I wanted to create a CI, but it took too long (more than an hour) to build.
Even if I make a CI, both Travis and Appveyor will fail.
I wanted to make conan-webrtc.... :(

Linux
```
#!/bin/bash
# from conanio/gcc7
# depot_tools need python2
pyenv uninstall -f $(pyenv versions --bare)
PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 2.7.14
pyenv global 2.7.14
export PATH=$PWD/depot_tools:$PATH
export PYTHONHTTPSVERIFY=0
/usr/bin/env python --version
#export GCLIENT_PY3=0

export GYP_GENERATORS=ninja
#export GYP_DEFINES=target_arch=x64
export GYP_DEFINES=component=shared_library target_arch=x64

git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
git clone https://webrtc.googlesource.com/src webrtc

mkdir webrtc-checkout
cd webrtc-checkout
fetch --nohooks webrtc
cd src
git checkout -b my_branch refs/remotes/branch-heads/4180
gclient sync --force

gn gen out/Debug --args='is_debug=true fatal_linker_warnings=false'
cd out/Debug
ninja
gn gen out/Release --args='is_debug=false fatal_linker_warnings=false'
cd out/Release
ninja
```

Windows
```
set "PATH=C:\Python27;C:\Python27\Scripts;%CD%\depot_tools;%PATH%"
set PYTHONHTTPSVERIFY=0
set DEPOT_TOOLS_WIN_TOOLCHAIN=0
set GYP_GENERATORS=ninja,msvs-ninja
set GYP_DEFINES=component=shared_library target_arch=x64

git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
git clone https://webrtc.googlesource.com/src webrtc

mkdir webrtc-checkout
cd webrtc-checkout
call fetch --nohooks webrtc
cd src
git checkout -b my_branch refs/remotes/branch-heads/4180
call gclient sync --force

call gn gen out/Debug --args="is_debug=true fatal_linker_warnings=false"
cd out/Debug
ninja
call gn gen out/Release --args="is_debug=false fatal_linker_warnings=false"
cd out/Release
ninja
```

**Comment by ZCube** [comment]
@fletcherdvalve Where do you use this code? Declared but not defined.

ice_session.obj : error LNK2019: g_fnWriteEvent_setsockopt
ice_session.obj : error LNK2019: g_fnWriteEvent_send
ice_session.obj : error LNK2019: g_fnWriteEvent_sendto

I think that solving this problem will probably solve three tasks first. Documentation is difficult to solve alone. Google Translator has its limitations.

**Comment by fletcherdvalve** [comment]
Whoa, thanks.  I can fix those linker errors, that was hacked in to a modified version of the WebRTC code.  It was to do some ETW tracing, when we weren't sure what WebRTC was doing.  But now we're trying to avoid modifying WebRTC at all, so we should just delete that.  I can look at that later, just delete the references to it or do what you need to do to move forward.

**Comment by ZCube** [comment]
@fletcherdvalve Finally I can build this library with steamnetworkingsockets_p2p_ice.cpp. How I can test this library? is there sample code ? or something guide ?

**Comment by ZCube** [comment]
Build instruction. webrtc license https://webrtc.org/support/license

```
git clone https://github.com/ZCube/GameNetworkingSockets
cd GameNetworkingSockets
mkdir webrtc
cd webrtc
curl -LO https://github.com/ZCube/GameNetworkingSockets/releases/download/test/webrtc_30a3e78794_header.7z
curl -LO https://github.com/ZCube/GameNetworkingSockets/releases/download/test/webrtc_30a3e78794_libs_debug.7z
curl -LO https://github.com/ZCube/GameNetworkingSockets/releases/download/test/webrtc_30a3e78794_libs_release.7z
7z x webrtc_30a3e78794_header.7z -oinclude/webrtc
7z x webrtc_30a3e78794_libs_release.7z -olib
cd ..
mkdir build
cd build
cmake -DUSE_MSVC_RUNTIME_DYNAMIC=ON -DUSE_STEAMWEBRTC=ON ..
```

**Comment by fletcherdvalve** [comment]
Wow, thanks again for doing all this work.

To test this, you need some sort of signaling service.  I'm not really sure how to do that in the context of this library.  If there is a standard / commonly used library for this, we could try to provide an example using that library or service.  I am not aware of one.  Am open to suggestions.

Thanks again.

**Comment by ZCube** [comment]
Then I will be waiting for the example. :)

**Comment by fletcherdvalve** [comment]
I could probably setup a dummy signaling service where you connect a TCP connection to the other end directly.  Just to be used as an example.  Obviously it you can actually connect a TCP connection directly, you don't need ICE or P2P, so it is totally just a fake test harness.  But it would server as an example of how to integrate a signaling service with the library.

**Comment by ZCube** [comment]
@fletcherdvalve 
example_p2p.cpp based console
https://gist.github.com/ZCube/306525ea25016ab05305a83e8e4307a9

```
➜  bin git:(117-webrtc) ✗ ./example_p2p
ip:::1
  0.000402 Connecting P2P socket to self (ip:::1).  Traffic will be relayed over the network
22BF0132750A310801122064CDA53DC3AFE9E4ED42CF15FD698366980ACBECB27679C7733D469C758DE956196BDE2D466806ADA9200928021240767FE80C3DDF89D72F0F23AB9772C6C067BAB04ADE4A1116C7349F2E8F2CE950DFE98F57E7001F698733501133EFBAC8529B63FDF5905B4496B7F8ADF6A79D0B3A4622440801122063DEA250A0FDFF27A18C3A1AEE2C056668437EA78918B595B1BDEDE7717C9DED50005A142212000000000000000000000000000000010000620669703A3A3A31420669703A3A3A314D256E5524520669703A3A3A315800
1
recieved
22BF0132750A3108011220937BA9F9BD9E1F6116DDED57DAD1934A52A396C3BD7C6583C620797DA28BB70B190B112A34A12464102009280212401ABEB62D160230E8338C3DE58C4C4B355B283F5C979577EF670F2B639C27983492EFAC03D1D6B3E5B33C51FC6461C361D72FA54FFCE6980B237CD2715F6403093A46224408011220B7BBB44196A65E2ACA32BBB40AADB2790B82057AC20E59FB4E9C860A590FB8C650005A142212000000000000000000000000000000010000620669703A3A3A31420669703A3A3A314DE83076BC520669703A3A3A315800
22BF0132750A3108011220937BA9F9BD9E1F6116DDED57DAD1934A52A396C3BD7C6583C620797DA28BB70B190B112A34A12464102009280212401ABEB62D160230E8338C3DE58C4C4B355B283F5C979577EF670F2B639C27983492EFAC03D1D6B3E5B33C51FC6461C361D72FA54FFCE6980B237CD2715F6403093A46224408011220B7BBB44196A65E2ACA32BBB40AADB2790B82057AC20E59FB4E9C860A590FB8C650005A142212000000000000000000000000000000010000620669703A3A3A31420669703A3A3A314DE83076BC520669703A3A3A315800
recieved
 16.014680 Ignoring P2P connection request from ip:::1.  We cannot accept it since we don't have a cert from Steam yet!
```
Any certificate needed ?

**Comment by fletcherdvalve** [comment]
Fixed 7b6bb5d17799407d4d57cf757887242136425719
